### ISSUE TITLE
Exclusive test, .only method

### ISSUE THREAD

**Issue by dy** [tapjs/tapjs#187]
Hi!
Is there a way to run any test exclusively, like [mocha](http://mochajs.org/#exclusive-tests)’s `it.only` or `describe.only`?
I guessed something like

``` js
t.test.only(function() {});
```

But there seems no that option.
That feature is quite useful in mocha, what is supposed way to quickly isolate single test in tap’s workflow?

**Comment by sindresorhus** [comment]
:+1: Should support multiple `test.only()` tests in contrast to Mocha which only allows one.

**Comment by isaacs** [comment]
This is tricky in node-tap.  Here's why:

``` javascript
var t = require('tap')
console.log('before')
t.pass('this is ok')
t.test('child', function (t) {
  t.plan(1)
  t.pass('child is ok')
})
console.log('mid')
t.test('child 2', function (t) {
  t.plan(1)
  t.pass('child 2 is ok')
})
console.log('after')
```

The output is this:

```
$ node t.js
before
TAP version 13
ok 1 - this is ok
    # Subtest: child
    1..1
    ok 1 - child is ok
ok 2 - child # time=6.092ms

mid
    # Subtest: child 2
    1..1
    ok 1 - child 2 is ok
ok 3 - child 2 # time=4.846ms

after
1..3
# time=41.233ms
```

In other words, `t.test` just runs right away, before moving on, so it can't know if there is another test coming up in the future with an `only` flag.  The reason for this is that I want `console.log()` calls to be in the "expected" place so it's easier to tie the output to the js test.

**Comment by rmg** [comment]
What _would_ be possible (maybe even trivial? famous last words..) to implement would be `t.skip()`..

Completely untested, but I'm imagining something like this:

``` js
Test.prototype.skip = function skip (name, extra, cb) { // matching .test()
  // some minimal effort to get a useful name
  if (typeof name === 'function') {
    name = name.name;
  }
  name = name || 'test'; 
  // this is basically what `{skip: true}` does when passed as the second arg to `test()`
  return this.pass('skipped: ' + name);
}
```

**Comment by isaacs** [comment]
@rmg I'd approach it like this:

``` JavaScript
Test.prototype.skip = function (name, extra, cb) {
  if (typeof name === 'function') {
    cb = name
    name = name.name || ''
    extra = {}
  } else if (typeof extra === 'function') {
    cb = extra
    extra = {}
  }

  if (!extra)
    extra = {}

  extra.skip = extra.skip || true
  return this.test(name, extra, cb)
}
```

But I'm not sure why `t.skip('child test', function (t) { ... })` is better than `t.test('child test', { skip: "reason" }, function (t) { ... })`.  I try to always have a reason why a test is skipped, so that it's clear what's going on.

Plus, if what you want is to run exactly one test, editing all the _other_ tests in the file to `s/test/skip/g` would be pretty rough.

My approach in this situation is to just move that test to the top of the file, and put a `return` after it.  Or better yet, split the tests into multiple files, and do `node test/foo.js` to just test that one thing.

**Comment by gr2m** [comment]
skip is possible today by passing `{skip: true}` to a test

``` js
t.test('check 1,2', {skip: true}, function (t) {
  t.plan(1)
  t.pass('child is ok')
})
```

I would really, really love an `.only` method though. Maybe check how https://github.com/substack/tape did it?

**Comment by dy** [comment]
@isaacs the issue reminds [delayed choice paradox in quantum mechanics](https://en.wikipedia.org/wiki/Delayed_choice_quantum_eraser), where the output result of the test can be altered after. One of possible solutions seems to be making all `t.test` asynchronous and capturing all the `console.log`’s and `stdout`’s to output them once at the final with filtered set of tests. Game isn’t worth the candles.

**Comment by isaacs** [comment]
Yeah, making t.test async (which is what tape does) means giving up consistent output (which is a source of problems), and hijacking stdout is fraught with unfathomable peril.  In fact, making t.test sync stripped away a TON of gnarly bugs in the initial release of node-tap, so I'd rather not go back there.

I think this is just a little too tricky, due to the fundamentally single-directional nature of time itself.  It's a reasonable request, but since it's probably not going to happen, I'm going to close this issue.  Sorry :(

**Comment by dy** [comment]
@isaacs actually there seems to be a simpler way to organize this.

First - run test with expectation of `test.only()` call, in child process, to avoid logging to the main output.
Second - if there is at least one `test.only` fired in the child process, just pipe output to the main process.
Third - if the child process didn’t call `test.only` at least once - invoke full test, as usual.
Something like:

``` js
var cp = require('child_process');
var child = cp.spawn('tap', ['./test.js', '--exclusive']); //or alike

//...obtaining the result of filtered test call through child.on('message', cb) or somehow

if (isOnlyInvoked) {
child.stdout.pipe(process.stdout);
}
else {
//...invoke the full test, the usual way
}
```

Anyways it is just an idea for technique, there are other options, like running test in child processes twice - with expectation of exclusive calls and without, and choose the best output.

**Comment by gr2m** [comment]
Thanks @isaacs, bummer, but well understandable

**Comment by isaacs** [comment]
@dfcreative Well, sure, but that just limits the _output_ -- the other tests are still actually run.  So, if they have side effects, take a long time, etc., then that doesn't help.

If you just want to control which tests get _output_, then that seems like a thing that could be handled by the output formatter, using a custom metadata field or something that the formatter could be aware of.  Doing a bunch of child proc juggling (and running the test file more than once!) seems like a lot of excessive complexity.

**Comment by mvoss9000** [comment]
I find tap and tape both special in their own way and I love them equally.  I do like the .only function in tape, and I wanted to create an enriched experience across both (satisfying my tabular ocd needs) - so I created [test-kit](https://github.com/quicbit-js/test-kit).  If you want .only plus ability to add functions across both, try it out.

**Comment by isaacs** [comment]
@mvoss9000 How do you maintain the constraint that synchronous tests are run synchronously?  Ie, this:

```js
var t = require('tap')
console.log('before')
t.test('child', function (t) {
  t.pass('this is fine')
  t.end()
})
console.log('after')
```

should output something like this:

```tap
before
TAP version 13
# Subtest: child
    ok 1 - this is fine
    1..1
ok 1 - child # time=5.099ms

after
1..1
```

The placement of `TAP version 13` is optional, of course, but the `before` and `after` should ideally be before and after the child test.

If you may encounter a `.only` somewhere along the test file, doesn't that mean you have to defer all test functions to at least the next tick?

What happens if you do something like this?

```js
t.test(function (t) { ... some stuff ... })
setTimeout(function () {
  t.test.only('only run this one but another one already ran oh noes!', function (t) { ... })
})
```

**Comment by mvoss9000** [comment]
Good point about console.log or other statements interspersed with the tests.  

I personally haven't realized the usefulness of synchronous relationship between code inside and outside of the tests. (and I would love to see the cases where it is important... really and sincerely mean that.  I'm not doubting that it is super useful, just saying I haven't seen the need yet in my testing.) 

Having the bodies of the tests themselves synchronous in relation to each other, has been all I needed so far:

So for the example above, I would do:

```javascript
var t = require('tap')
t.test('child', function (t) {
  console.log('before')
  t.pass('this is fine')
  t.end()
  console.log('after')
})
```

But I would be happy to enrich test-kit to allow synchronous statements outside the tests:

```javascript
t.log('before')
t.test('child', function (t) {
  t.pass('this is fine')
  t.end()
})
t.log('after')
```

... where <code>t.log(..)</code> could just be a special case of a general hook we could make:

    t.sync(function() {...})

That executes statements sequentially/synchronously with respect to the tests.  (I agree very much with your earlier point that messing with console.log is a *bad* idea).

As for the point about setTimeout()... I'm not sure I would worry about that case because the cost of solving it exceeds the handiness of .only(), which is just there to save time when I want to quickly hone in on a test.  

But if we wanted a way to hone in on certain tests, we could add another method to t that turns tests on and off by their name and/or number:

t.only(1), would run only the first test.  In fact, I might want to put a statement like this at the top of the test file and eliminate all the async stuff in test-kit (make it opt-in).... hadn't thought of that until just now.  I mean it's nice to switch tests on and off right at the test, but putting the statement early in the module might be 'good enough'.

**Comment by mvoss9000** [comment]
... Thinking about this a bit.  Having only() take an expression, number or string, rather than a test function has some advantages.

```javascript
var t = require('tap')
t.only(/child/)        // only run tests with child in the name
t.only('child')        // only run 'child' test
t.only(/^child$/)      // only run 'child' test 
t.only(3)              // only run test #3  (child test, in this case)
t.only(2,3)            // only run tests 2 and 3 (two through 3)

t.test('test one', function(t) {
  ...
})

t.test('test two', function(t) {
  ...
})
...
console.log('before')
t.test('child', function (t) {
  t.pass('this is fine')
  t.end()
})
console.log('after')

```

The .only() call should output information about the filter being applied (so it isn't left on by accident)

    # ONLY running tests matching: /child/
    ...

**Pros**

* async isn't needed.  execution order remains simple.
* test selection can be switched on and off by commenting out t.only() statements, which is sometimes convenient.

**Cons**

* the t.only() statement has to be placed above any test calls, so users can't select a test right at the point of work (in the middle of a long test file) 

For test-kit, I could implement this benign '.only()' as an option when using either tap or tape, and add a opt-in switch for hard-core .only() users who really really want it in both tape and tap.  That way, tap stays synchronous/normal by default.

**Comment by mvoss9000** [comment]
Any thoughts on the .only() expressions?  

There are actually more important features in [test-kit](https://github.com/quicbit-js/test-kit) that I would be happy to see folded into the mainstream tap tool.  Test tables and descriptive messages about input/output are quite helpful in my experience.  I know everyone's busy, but you might find these features help to shorten and clarify tests.

**Comment by isaacs** [comment]
I'm not really interested in changing the user-facing API that much.  `t.test()` stays synchronous.

An `only()` method that takes a filter expression is sort of nice, but it does unfortunately mean that you have to set those things at the top of the test file while developing.  I also don't love variadic arg types if we can get away with it.  And the spelling is confusing with mocha's `describe.only()` and ava's `t.test.only()`.

I'd have the runner just mark the skipped tests as if the user had done `t.test('name', { skip: true })`

Re adding asserts and stuff, do you know about [t.addAssert](http://www.node-tap.org/advanced/#taddassertname-length-fn)?

**Comment by mvoss9000** [comment]
Understand about only().  Case closed.  

I did see t.addAssert... *after I wrote the first cut* of test-kit.  Great feature.  I was then going to use it, but then realized that for test-kit, which decorates tape/node-tap/... or other lib in the same way, it would just add another code path.  But I have in mind to add it for tap if the sloppy decoration gets slow or fails in some way.

But I think the tableAssert stuff is more important than only() (perhaps it is rude to add this in this thread?  This is my first communication experience on open source, so I'm not sure of etiquette)

[qb-uint/test](https://github.com/quicbit-js/qb-uint/blob/master/test.js)

```javascript
test('uint: throws', function(t) {
    var buf = new Uint8Array(5)
    t.tableAssert(
        [
            [ 'fn', 'input',  'exp' ],
            [ 'size', [Math.pow(2, 53)],    /greater than max safe integer/ ],
            [ 'size', [1/0],                /greater than max safe integer/ ],
            [ 'size', [-1],                 /cannot be negative/ ],
            [ 'size', [-1/0],               /cannot be negative/ ],
            [ 'size', [],                   /cannot convert/    ],
            [ 'write', [0x7FFFFFFF+1, buf], /too large/         ],
        ],
        function(fn, input) {
            uint[fn].apply(null, input)
        },
        { assert: 'throws' }
    )
})
```
... which gives clearer output:

```
TAP version 13
# uint: throws
ok 1 : ('size',[9007199254740992]) -expect-> ('/greater than max safe integer/')
ok 2 : ('size',[null]) -expect-> ('/greater than max safe integer/')
ok 3 : ('size',[-1]) -expect-> ('/cannot be negative/')
ok 4 : ('size',[null]) -expect-> ('/cannot be negative/')
ok 5 : ('size',[]) -expect-> ('/cannot convert/')
ok 6 : ('write',[2147483648,{'0':0,'1':0,'2':0,'3':0,'4':0}]) -expect-> ('/too large/')
```

**Comment by gr2m** [comment]
@isaacs I wonder if it would be possible to run `tap` in a dedicated only mode, in which it would only execute tests with a `{only: true}` to it, basically the opposite of `{skip: true}`. By default, `{only: true}` would be ignored, but running with `tap --only` or something similar would be as if I added `{skip: true}` to all over tests. Would something like that be possible?

**Comment by isaacs** [comment]
Huh.  @gr2m that's a pretty good idea.

While we're at it, maybe we could add a way to say "run the skipped tests, too", so you could mark things as skipped in general, but run them while you're developing?  Or skip really slow tests, but still run them on CI?

**Comment by isaacs** [comment]
I guess you can already do stuff like `{ skip: process.env.CI ? 'skip in ci' : false }`, so maybe "ignore skip" isn't such a valuable addition.

**Comment by gr2m** [comment]
I’d be happy to try send a pull request for the `--only` CLI option if you like? I could use some pointers on where the code would need to be touched, how test and where to update documentation. Just a bit to get started, we can then discuss details in the pull request itself. 

This would be tremendously helpful to have in `tap` :) I also like that it would allow to run multiple tests with `only` option but without the risk to accidentally push a `test.only` change which would run the single test in CI, this bit me lots of time in the past. I think the `{only: true}` option with `--only` has less potential for developers to shoot themselves in the foot

**Comment by isaacs** [comment]
@gr2m There are a few parts to this, here's how I'd tackle it:

1. (Maybe only on the TAP object?) Have some way to say "only run child tests that have `only` in their options".  Set this by default if `TAP_ONLY=1` is set in the environment.
2. In the runner, check for a `--only` flag, and set that option in the environment for the child processes it spawns.
3. Add a `t.only(...)` command that calls out to `t.test()` but sets `only: true` in the options.  If the runner doesn't have that `only` option set, then maybe print a warning to stderr so the user knows that they should be running with `tap --only` or something?

**Comment by gr2m** [comment]
> Add a t.only(...) command

Are you sure you want to add the method? There is no `t.skip` is there? I’d suggest we add the `{only: true}` option first and leave adding a `t.only` and maybe a `t.skip` method to a follow up issue?

**Comment by isaacs** [comment]
Yeah, that's a good approach.

**Comment by isaacs** [comment]
I was thinking it'd be nice to maybe implement this as a filter function on the root TAP object.

Then it can also be leveraged to do `tap --filter="some pattern"` to filter tests based on their name.

**Comment by gr2m** [comment]
@isaacs I started a work-in-progress pull request at https://github.com/tapjs/node-tap/pull/374. Could you have a look if that goes in the right direction? It has no tests yet, I want to make sure I get the implementation right first.

If it’s looking good, I’d love to hear how you would approaching testing it

**Comment by gr2m** [comment]
released in 10.7 via c41742440869f06ee1eac39d9fc80a3fc3af3f53
